### PR TITLE
util: fixed memory leaks in zenfs utility

### PR DESCRIPTION
**Pull request by percona-ysorokin**
Fixed leak in 'ZenFS::MkFS()' caused by a 'Superblock' allocated on
heap and never deallocated.

Fixed resource leak in 'ListZenFileSystems()' where 'opendir()' did not
have corresponding 'closedir()'

Reworked 'zbd_open()' and 'zenfs_mount()' functions so that they now
accept/return 'std::unique_ptr's for 'ZonedBlockDevice' and 'ZenFS'
instead of raw pointers. This fixed a number of memory leaks in
'zenfs' utility command handlers.

Fixed 'std::string's passed by value instead of by const reference in
some function signatures.

Signed-off-by: Yura Sorokin <yura.sorokin@percona.com>

### CODE REVIEW

File: util/zenfs.cc [hunk 1]
```diff
@@ -36,36 +37,39 @@ DEFINE_string(backup_path, "", "Path to backup files");
 
 namespace ROCKSDB_NAMESPACE {
 
-ZonedBlockDevice *zbd_open(bool readonly, bool exclusive) {
-  ZonedBlockDevice *zbd = new ZonedBlockDevice(FLAGS_zbd, nullptr);
+using ZonedBlockDevicePtr = std::unique_ptr<ZonedBlockDevice>;
```
Comment (yhr) [inline util/zenfs.cc:40]: Please don't introduce new types unless it is really needed, we want to stick with rocksdb coding style (and I think adding types just makes the code harder to read - just another level of indirection)

File: util/zenfs.cc [hunk 2]
```diff
@@ -79,41 +83,40 @@ int zenfs_tool_mkfs() {
     return 1;
   }
 
-  ZonedBlockDevice *zbd = zbd_open(false, true);
-  if (zbd == nullptr) return 1;
+  auto zbd = zbd_open(false, true);
```
Comment (yhr) [inline util/zenfs.cc:86]: The only place I want auto to be used is for iterators in for loops - I'm very much in favor of typed languages :)

File: util/zenfs.cc [hunk 3]
```diff
@@ -79,41 +83,40 @@ int zenfs_tool_mkfs() {
...
         "Existing filesystem found, use --force if you want to replace it.\n");
     return 1;
   }
 
-  if (zenFS != nullptr) delete zenFS;
+  zenFS.reset();
 
   zbd = zbd_open(false, true);
-  zenFS = new ZenFS(zbd, FileSystem::Default(), nullptr);
+  auto *zbdRaw = zbd.get();
```
Comment (yhr) [inline util/zenfs.cc:101]: See prev comment

File: util/zenfs.cc [hunk 4]
```diff
@@ -167,12 +169,11 @@ void format_path(std::string &path)
...
-  Status s;
-  ZonedBlockDevice *zbd = zbd_open(true, false);
-  if (zbd == nullptr) return 1;
+  auto zbd = zbd_open(true, false);
+  if (!zbd) return 1;
 
-  ZenFS *zenFS;
-  s = zenfs_mount(zbd, &zenFS, true);
+  ZenFSPtr zenFS;
+  auto s = zenfs_mount(zbd, &zenFS, true);
```
Comment (yhr) [inline util/zenfs.cc:176]: Same here, no need to use auto

File: util/zenfs.cc [hunk 5]
```diff
@@ -186,20 +187,20 @@ int zenfs_tool_list() {
...
-  ZonedBlockDevice *zbd = zbd_open(true, false);
-  if (zbd == nullptr) return 1;
+  auto zbd = zbd_open(true, false);
+  if (!zbd) return 1;
+  auto *zbdRaw = zbd.get();
 
-  ZenFS *zenFS;
-  s = zenfs_mount(zbd, &zenFS, true);
+  ZenFSPtr zenFS;
+  auto s = zenfs_mount(zbd, &zenFS, true);
```
Comment (yhr) [inline util/zenfs.cc:195]: Please use full type declaration

File: util/zenfs.cc [hunk 6]
```diff
@@ -212,18 +213,17 @@ int zenfs_tool_df() {
 }
 
 int zenfs_tool_lsuuid() {
-  std::map<std::string, std::string>::iterator it;
-  std::map<std::string, std::string> zenFileSystems = ListZenFileSystems();
+  auto zenFileSystems = ListZenFileSystems();
 
-  for (it = zenFileSystems.begin(); it != zenFileSystems.end(); it++)
-    fprintf(stdout, "%s\t%s\n", it->first.c_str(), it->second.c_str());
+  for (const auto &p: zenFileSystems)
```
Comment (yhr) [inline util/zenfs.cc:218]: here auto makes sense

File: util/zenfs.cc [hunk 7]
```diff
@@ -274,56 +274,52 @@ IOStatus zenfs_tool_copy_file(FileSystem *f_fs, std::string f, FileSystem *t_fs,
...
   s = t_fs->NewWritableFile(t, fopts, &t_file, &dbg);
   if (!s.ok()) { return s; }
 
   t_file->SetWriteLifeTimeHint(GetWriteLifeTimeHint(t));
 
-  buffer = (char *)malloc(buffer_sz);
-  if (buffer == nullptr) {
-    return IOStatus::IOError("Failed to allocate copy buffer");
-  }
+  buffer.resize(buffer_sz);
```
Comment (yhr) [inline util/zenfs.cc:292]: It's nice to reduce the number of lines, but an allocation failure can still happen, right?

File: util/zenfs.cc [hunk 8]
```diff
@@ -356,36 +352,34 @@ IOStatus zenfs_tool_copy_dir(FileSystem *f_fs, std::string f_dir, FileSystem *t_
 }
 
 int zenfs_tool_backup() {
-  Status status;
   IOStatus io_status;
-  ZonedBlockDevice *zbd;
-  ZenFS *zenFS;
+  auto zbd = zbd_open(true, true);
```
Comment (yhr) [inline util/zenfs.cc:356]: No auto declarations please

File: util/zenfs.cc [hunk 9]
```diff
@@ -410,17 +401,18 @@ int zenfs_tool_restore() {
 
   ReadWriteLifeTimeHints();
 
-  zbd = zbd_open(false, true);
-  if (zbd == nullptr) return 1;
+  auto zbd = zbd_open(false, true);
```
Comment (yhr) [inline util/zenfs.cc:404]: No autos

File: util/zenfs.cc [hunk 10]
```diff
@@ -431,12 +423,12 @@ int zenfs_tool_restore() {
...
-  if (zbd == nullptr) return 1;
+  auto zbd = zbd_open(true, false);
+  if (!zbd) return 1;
+  auto *zbdRaw = zbd.get();
 
-  ZenFS *zenFS;
-  s = zenfs_mount(zbd, &zenFS, true);
+  ZenFSPtr zenFS;
+  auto s = zenfs_mount(zbd, &zenFS, true);
   if (!s.ok()) {
```
Comment (yhr) [inline util/zenfs.cc:432]: Same here, please write out the type in stead

**Review by yhr — CHANGES_REQUESTED** [review]
Thanks for the PR, looks good over all - but there are some nitpicks (mostly coding style) that needs to be addressed.

It would have been easier to review if the different fixes had been split up in individual patches, please try to do that next time. Small, one logical change per commit, patches helps when bisecting for regressions as well.

**Comment by yhr** [comment]
zenfs integration testing ((util, smoke, quick performance test sets) passed

**Comment by percona-ysorokin** [comment]
@yhr I addressed your comments:
- removed `ZonedBlockDevicePtr` and `ZenFSPtr` typedefs.
- `auto`s changed to explicit types
- removed formatting fixes
- reimplemented OOM check after buffer allocation in `zenfs_tool_copy_file()`
- removed utility help line fix (will be submitted as a separate PR)

**Review by yhr — APPROVED** [review]

**Comment by yhr** [comment]
Looks good to me and passed integration testing

**Pull request merged**